### PR TITLE
Adapter cache service

### LINKED ISSUE #97: Very slow data extraction through opendap when using twitcher
Working on a notebook running on _boréas_ I noticed that using twitcher really slows down the process.

MWE:
```python3
import xarray as xr

def extract(url):
    ds = xr.open_dataset(url, drop_variables=['ts', 'time_vectors'],
                         chunks={'time': 256, 'lat': 16, 'lon': 16})
    ds.isel(lon=slice(0, 10), lat=slice(0, 10)).to_netcdf('test.nc', compute=True)
```

### Test 1 using twitcher (normal url)
```python3
%%time
baseurl = "http://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/testdata/ouranos/cb-oura-1.0_rechunk/MPI-ESM-LR/rcp85/day/MPI-ESM-LR_rcp85_allvars.ncml"
extract(baseurl)
```
Output:
 ```
CPU times: user 4.51 s, sys: 1.79 s, total: 6.3 s
Wall time: 3min 10s
```
And CPU usage (of the notebook) is super low, around 2-3 % for most of the 3 mn. I can see that twitcher (on boréas) runs with 30-60% CPU.

### Test 2 bypassing twitcher:
```python3
%%time
baseurl = "http://pavics.ouranos.ca:8083/twitcher/ows/proxy/thredds/dodsC/birdhouse/testdata/ouranos/cb-oura-1.0_rechunk/MPI-ESM-LR/rcp85/day/MPI-ESM-LR_rcp85_allvars.ncml"
extract(baseurl)
```
Output:
```
CPU times: user 3.56 s, sys: 996 ms, total: 4.56 s
Wall time: 11.3 s
```
Also, CPU usage of the notebook stays reasonable (30-40%).

This is a simple example with a small region, but in my notebook I need to extract the same region (potentially large) from 23 different datasets.  Going through twitcher means I have to wait around 1 hr in some cases, even when launching multiple (2-3) extraction in parallel.­..

### PR DESCRIPTION
**Pull request by fmigneault**
helps resolve bird-house/twitcher#97 
based on finding from https://github.com/bird-house/birdhouse-deploy/compare/cache-settings
and https://github.com/bird-house/twitcher/issues/97#issuecomment-780747482 

settings should be defined in `Twitcher` 's INI with something like: 

```
# caching settings for specific sections/functions
cache.regions = acl, service
cache.type = memory
cache.acl.expire = 60
cache.acl.enabled = true
cache.service.expire = 10
cache.service.enabled = true
```

### CODE REVIEW

File: magpie/adapter/magpieservice.py [hunk 1]
```diff
@@ -68,30 +72,60 @@ def list_services(self, request=None):  # noqa: F811
         json_body = resp.json()
         for service_type in json_body["services"]:
             for service in json_body["services"][service_type].values():
-                services.append(Service(url=service["service_url"],
-                                        name=service["service_name"],
-                                        type=service["service_type"]))
+                services.append(TwithcerService(url=service["service_url"],
```
Comment (dbyrns) [inline magpie/adapter/magpieservice.py:75]: Typo in class name `TwithcerService`

**Review by dbyrns — CHANGES_REQUESTED** [review]
A little typo but everything else is good!

**Comment by dbyrns** [comment]
Could you also contribute to https://github.com/bird-house/birdhouse-deploy/compare/cache-settings so that the `service` region is also properly configured?

File: magpie/adapter/magpieservice.py (hunk 1 above)
Comment (fmigneault) [reply magpie/adapter/magpieservice.py:75]: Well spotted :)

**Comment by fmigneault** [comment]
> Could you also contribute to https://github.com/bird-house/birdhouse-deploy/compare/cache-settings so that the `service` region is also properly configured?

For the moment, that would give the wrong indication that caching is defined for services but it won't be working until Magpie 3.7 is integrated.

**Comment by dbyrns** [comment]
Ok but we definitely not want to forget that piece of config when we will upgrade Magpie to 3.7

**Review by dbyrns — DISMISSED** [review]
💯

**Comment by ldperron** [comment]
> Ok but we definitely not want to forget that piece of config when we will upgrade Magpie to 3.7

We can already do the version bump and the config change in https://github.com/bird-house/birdhouse-deploy/compare/cache-settings , and issue the PR when 3.7 is ready.

**Comment by fmigneault** [comment]
> 
> We can already do the version bump and the config change in https://github.com/bird-house/birdhouse-deploy/compare/cache-settings , and issue the PR when 3.7 is ready.

@ldperron 
Magpie has been sticking to 1.7.3 in birdhouse-deploy for way too long. 

Can be adjusted following https://github.com/bird-house/birdhouse-deploy/pull/102 and https://github.com/bird-house/birdhouse-deploy/pull/107 tests are passing, but everything seems to work for me with "3.7-dev".

@dbyrns 
> 
> Ok but we definitely not want to forget that piece of config when we will upgrade Magpie to 3.7

I will add it (doesn't hurt), but will also put a note as it won't take effect (contrary to `acl` one).

**Review by tlvu — DISMISSED** [review]
Approve in principle only since I can not pull this version into PAVICS stack and test it.

**Comment by tlvu** [comment]
> Can be adjusted following [bird-house/birdhouse-deploy#102](https://github.com/bird-house/birdhouse-deploy/pull/102) and [bird-house/birdhouse-deploy#107](https://github.com/bird-house/birdhouse-deploy/pull/107) tests are passing, but everything seems to work for me with "3.7-dev".

@fmigneault Something I meant to ask you but never got around to do it is why 2 separate PR to bump magpie in PAVICS stack?  Is one supposed to replace the other one?   At the end of the day, I suppose both will have the latest and greatest magpie with all the fixes so why 2 separate PR?

From the review perspective, it's confusing for me and double my review work.  I am not sure which one I need to prioritize.  You also added a notebook in the PAVICS-e2e-workflow-tests and I ended up asking you (https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/55#issuecomment-776772512) which of the 2 magpie PR the notebook was intended for.

**Comment by fmigneault** [comment]
@tlvu 

> why 2 separate PR to bump magpie in PAVICS stack? Is one supposed to replace the other one? At the end of the day, I suppose both will have the latest and greatest magpie with all the fixes so why 2 separate PR?

Simply because I was not the author of bird-house/birdhouse-deploy#102 and I add even more features in bird-house/birdhouse-deploy#107 than was the first provides.
PR bird-house/birdhouse-deploy#107 is to be merged in bird-house/birdhouse-deploy#102, and the later in master (or whichever combination, as long as latest magpie is picked). 
Tests are against bird-house/birdhouse-deploy#107

**Comment by tlvu** [comment]
> > 
> 
> Simply because I was not the author of [bird-house/birdhouse-deploy#102](https://github.com/bird-house/birdhouse-deploy/pull/102) and I add even more features in [bird-house/birdhouse-deploy#107](https://github.com/bird-house/birdhouse-deploy/pull/107) than was the first provides.
> PR [bird-house/birdhouse-deploy#107](https://github.com/bird-house/birdhouse-deploy/pull/107) is to be merged in [bird-house/birdhouse-deploy#102](https://github.com/bird-house/birdhouse-deploy/pull/102), and the later in master (or whichever combination, as long as latest magpie is picked).
> Tests are against [bird-house/birdhouse-deploy#107](https://github.com/bird-house/birdhouse-deploy/pull/107)

Oh right, those 2 PR opened for too long I forgot they are related and not separate PR.

So let's not add more "feature" to those 2 PR, only required fixes, get Jenkins test-suite to pass to ensure no regressions and merge those 2 PR.

There is also a Weaver PR opened.  I am guess we will focus on closing the Magpie first?

**Comment by fmigneault** [comment]
@tlvu 
Magpie PRs are blocked mostly by other tests failing the test suite.
On one hand, we add security, but by adding it, other things that assumed open-bar data access breaks. 

As for Weaver one, it would be coming soon, when tests of https://github.com/crim-ca/weaver/pull/214 pass.

**Pull request merged**